### PR TITLE
8303689: javac -Xlint could/should report on "dangling" doc comments

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review the updates to support a proposed new `-Xlint:dangling-doc-comments` option.

The work can be thought of as in 3 parts:

1. An update to the `javac` internal class `DeferredLintHandler` so that it is possible to specify the appropriately configured `Lint` object when it is time to consider whether to generate the diagnostic or not.

2. Updates to the `javac` front end to record "dangling docs comments" found near the beginning of a declaration, and to report them using an instance of `DeferredLintHandler`. This allows the warnings to be enabled or disabled using the standard mechanisms for `-Xlint` and `@SuppressWarnings`.  The procedure for handling dangling doc comments is described in this comment in `JavacParser`.
```
     *  Dangling documentation comments are handled as follows.
     *  1. {@code Scanner} adds all doc comments to a queue of
     *     recent doc comments. The queue is flushed whenever
     *     it is known that the recent doc comments should be
     *     ignored and should not cause any warnings.
     *  2. The primary documentation comment is the one obtained
     *     from the first token of any declaration.
     *     (using {@code token.getDocComment()}.
     *  3. At the end of the "signature" of the declaration
     *     (that is, before any initialization or body for the
     *     declaration) any other "recent" comments are saved
     *     in a map using the primary comment as a key,
     *     using this method, {@code saveDanglingComments}.
     *  4. When the tree node for the declaration is finally
     *     available, and the primary comment, if any,
     *     is "attached", (in {@link #attach}) any related
     *     dangling comments are also attached to the tree node
     *     by registering them using the {@link #deferredLintHandler}.
     *  5. (Later) Warnings may be genereated for the dangling
     *     comments, subject to the {@code -Xlint} and
     *     {@code @SuppressWarnings}.
```

3.  Updates to the make files to disable the warnings in modules for which the 
warning is generated.  This is often because of the confusing use of `/**` to
create box or other standout comments.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8329137](https://bugs.openjdk.org/browse/JDK-8329137) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8303689](https://bugs.openjdk.org/browse/JDK-8303689): javac -Xlint could/should report on "dangling" doc comments (**Enhancement** - P4)
 * [JDK-8329137](https://bugs.openjdk.org/browse/JDK-8329137): javac -Xlint could/should report on "dangling" doc comments (**CSR**)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [39689a52](https://git.openjdk.org/jdk/pull/18527/files/39689a5255ce4e9a4333f9bb0f01995a86037192)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to [3d6f1f95](https://git.openjdk.org/jdk/pull/18527/files/3d6f1f95ed6df0575d7550337869f7adf5af18cb)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18527/head:pull/18527` \
`$ git checkout pull/18527`

Update a local copy of the PR: \
`$ git checkout pull/18527` \
`$ git pull https://git.openjdk.org/jdk.git pull/18527/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18527`

View PR using the GUI difftool: \
`$ git pr show -t 18527`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18527.diff">https://git.openjdk.org/jdk/pull/18527.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18527#issuecomment-2024074849)

### CODE REVIEW

**Comment by pavelrappo** [comment]
Would this be the first lint -- not doclint -- warning related to comments, let alone doc comments?

**Comment by pavelrappo** [comment]
Javadoc changes look trivially good. I only note that the javadoc man page diff contains some unrelated changes.

File: src/jdk.javadoc/share/man/javadoc.1 [hunk 1]
```diff
@@ -96,6 +96,22 @@ option either to recursively traverse a directory and its
...
+immediately precedes the beginning of the declaration, whether that is
+an annotation, modifier, or the name being declared.
+If there are multiple documentation comments before the declaration,
+only the last one (closest to the declaration) will be used.
+If there are any documentation comments after the beginning of the
+declaration, they will be ignored.
+To check for any extra or misplaced documentation comments, compile your
+source code with the \f[V]javac\f[R] option \f[V]-Xlint\f[R], or more
+specifically, \f[V]-Xlint:dangling-doc-comments\f[R].
+Within a source file, you may use suppress any warnings generated by
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/man/javadoc.1:111]: Typo?
```suggestion
Within a source file, you may suppress any warnings generated by
```

Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/man/javadoc.1:111]: Thanks; I'll check the underlying original.

**Comment by jonathan-gibbons** [comment]
> Would this be the first lint -- not doclint -- warning related to comments, let alone doc comments?

No. `-Xlint:dep-ann` correlates `@Deprecated` annotations with `@deprecated` tags in doc comments.

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

**Review by magicus — APPROVED** [review]
The build changes look okay. 

Do you have any plan of going through all the Java modules and fixing the issues, or opening JBS issues to have them fixed? Or will these lint warnings remain disabled for the foreseeable future?

**Comment by jonathan-gibbons** [comment]
@magicus 
> The build changes look okay.
> 
> Do you have any plan of going through all the Java modules and fixing the issues, or opening JBS issues to have them fixed? Or will these lint warnings remain disabled for the foreseeable future?

The plan is to create an umbrella bug to clean up the individual modules. There is interest to clean up `java.base`, to keep that one free of any warnings, and I can see that the lang tools modules will get cleaner up as well. It will be up to other component teams to decide if and when to clean up other parts of the system.  Once this work has been integrated, it is relatively easy to enable the warnings for a module and to fix the ensuing issues. Since any changes "only" involve comments, it should be reasonably easy to fix them, unlike some pervasive other warnings, like `this-escape`.

**Review by vicente-romero-oracle — APPROVED** [review]
looks good

**Review by prrace — APPROVED** [review]

**Pull request closed without merging**